libs/langchain-core/src/_standard/message.ts

```diff
@@ -0,0 +1,512 @@
+ *   };
+ *   properties: {
+ *     ai: { model: string };
+ *   }
+ * }
+ *
+ * // Merged structure allows both text and images in human messages
+ * // AI messages have both confidence and model properties
+ * type Merged = $MergeMessageStructure<StructureA, StructureB>;
+ * ```
+ *
+ * @template A - First message structure to merge
+ * @template B - Second message structure to merge (takes precedence over A)
+ */
+export interface $MergeMessageStructure<
+  A extends $MessageStructure,
+  B extends $MessageStructure
+> {
+  tools: $Merge<A["tools"], B["tools"]>;
+  contentBlocks: {
+    [K in keyof (A["contentBlocks"] &
+      B["contentBlocks"])]: K extends keyof A["contentBlocks"] &
+      keyof B["contentBlocks"]
+      ? $MergeDiscriminatedUnion<
+          NonNullable<A["contentBlocks"][K]> & Record<"type", PropertyKey>,
+          NonNullable<B["contentBlocks"][K]> & Record<"type", PropertyKey>,
+          "type"
+        >
+      : K extends keyof A["contentBlocks"]
+      ? A["contentBlocks"][K]
+      : K extends keyof B["contentBlocks"]
+      ? B["contentBlocks"][K]
+      : never;
+  };
+  properties: {
+    [K in keyof (A["properties"] &
+      B["properties"])]: K extends keyof A["properties"] & keyof B["properties"]
+      ? $Merge<NonNullable<A["properties"]>[K], NonNullable<B["properties"]>[K]>
+      : K extends keyof A["properties"]
+      ? A["properties"][K]
+      : K extends keyof B["properties"]
+      ? B["properties"][K]
+      : never;
+  };
+}
+
+/**
+ * Gets all possible message types from a message structure
+ *
+ * This is used internally to get the possible message types for a given message structure.
+ *
+ * If a message structure has no tools, it will not include the "tool" type. If a message structure
+ * has arbitrary types defined in either contentBlocks or properties, it will include those types
+ * as available message types.
+ *
+ * @template TStructure - The message structure to get types from
+ * @example
+ * ```ts
+ * interface MyStructure extends $MessageStructure {
+ *   contentBlocks: { human: ContentBlock.Text };
+ *   properties: { ai: { confidence: number } };
+ *   tools: Tools.ToolCall;
+ * }
+ *
+ * type Types = $MessageStructureTypes<MyStructure>;
+ * // Types = "human" | "ai" | "tool"
+ * ```
+ */
+export type $MessageStructureTypes<TStructure extends $MessageStructure> =
+  | keyof TStructure["contentBlocks"]
+  | keyof TStructure["properties"]
+  | (TStructure["tools"] extends undefined ? never : "tool");
+
+/**
+ * Standard message structured used to define the most basic message structure that's
+ * used throughout the library.
+ *
+ * This is also the message structure that's used when a message structure is not provided.
+ */
+export interface $StandardMessageStructure extends $MessageStructure {
+  /** @internal Discriminator to give TS a hint when evaluating if a type is a standard message structure */
+  __STANDARD_MESSAGE_STRUCTURE: never;
```

Comment:
Should we use a symbol instead? TypeScript will still discriminate properly but it won't be shown in Intellisense.

Alternatively, would use `~` instead of `__` to avoid bubbling it upwards?